pkg/common/cns-lib/volume/taskmanager.go

```diff
@@ -0,0 +1,356 @@
+	MarkedForRemoval bool
+	ResultCh         chan Result
+}
+
+// RemoveTasksMarkedForDeletion goes over the list of tasks in the map
+// and removes tasks that have been marked for deletion
+func (t *TaskViewManagerImpl) RemoveTasksMarkedForDeletion() {
+	ctx := context.Background()
+	log := logger.GetLogger(ctx)
+	listView := t.getListView()
+	if listView == nil {
+		log.Errorf("ListView is empty. Will attempt to remove invalid tasks in next attempt. ")
+		return
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	log.Debugf("pending tasks count before purging: %v", len(t.taskMap))
+	for taskMoRef, taskDetails := range t.taskMap {
+		if taskDetails == nil {
+			log.Errorf("failed to remove task %v from ListView as taskDetails is nil", taskMoRef)
+		}
+		if taskDetails.MarkedForRemoval == true {
+			err := listView.Remove(ctx, []types.ManagedObjectReference{taskMoRef})
+			if err != nil {
+				log.Errorf("failed to remove task from ListView. error: %+v", err)
+				continue
+			}
+			delete(t.taskMap, taskMoRef)
+		}
+	}
+	log.Debugf("pending tasks count after purging: %v", len(t.taskMap))
+}
+
+// MarkForDeletion marks a given task MoRef for deletion by setting a boolean flag in the TaskDetails object
+func (t *TaskViewManagerImpl) MarkTaskForDeletion(ctx context.Context, taskMoRef types.ManagedObjectReference) error {
+	log := logger.GetLogger(ctx)
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// ch, ok := t.taskMap[TaskDetails{Reference: taskMoRef}]
+	taskDetails, ok := t.taskMap[taskMoRef]
+	if !ok || taskDetails == nil {
+		return logger.LogNewErrorf(log, "failed to retrieve taskDetails for %+v", taskMoRef)
+	}
+	// delete(t.taskMap, taskMoRef)
+	taskDetails.MarkedForRemoval = true
+	return nil
+}
+
+func (t *TaskViewManagerImpl) GetPendingTasksCount() int {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	return len(t.taskMap)
+}
+
+func NewTaskViewManager(virtualCenter *cnsvsphere.VirtualCenter) *TaskViewManagerImpl {
+	ctx := context.Background()
+	m := make(map[types.ManagedObjectReference]*TaskDetails)
+	t := &TaskViewManagerImpl{
+		virtualCenter: virtualCenter,
+		ctx:           ctx,
+		taskMap:       m,
+	}
+	go t.init(t.ctx)
+	return t
+}
+
+// TaskViewManagerImpl is the struct used to manage a single listView instance.
+// params
+// virtualCenter: holds a reference to the global VC object
+// sync.Mutex: is required to handle race conditions for
+// storing and retrieving tasks during a driver restart
+// taskMap: is an in-memory map to associate task details
+// to the channel created by the caller to receive the task result
+// context.Context: is separate from the context set by CSI ops
+// listView: holds the managed object used to monitor multiple concurrent VC tasks
+// shouldStopListening: in case of regular CSI operation, even after receiving a batch of updates,
+// we want to continue listening for subsequent updates.
+// in case of unit tests, we need to stop listening for the test to complete execution
+type TaskViewManagerImpl struct {
```

Comment:
Can you move this struct up in this file and then start interface implementation?

Reply:
Fixed.